### ISSUE TITLE
Apisix & Gateway API is not working.

### ISSUE THREAD

**Issue by MBoui** [OpenFunction/OpenFunction#358]
**Description**
I am trying to configure openfunction working with apisix and gateway api. I was able to create a simple route with the echo server to check if apisix & gateway api is working.

Now i am stuck with OpenFunction.

**Environmental**

- Operating System Information: ubuntu
- kubernetes version: lastest (with minikube)
- OpenFunction version: lastest

**Expected behavior**
Request the function

**Actual behavior**
Get this error

```
curl -i $SERVICE_IP/v2/foo/world -H 'Host: sample.ofn.io'
HTTP/1.1 404 Not Found
Content-Type: text/plain; charset=utf-8
Content-Length: 73
Connection: keep-alive
X-Content-Type-Options: nosniff
Date: Tue, 30 Aug 2022 21:41:02 GMT
Server: APISIX/2.15.0

Error getting active endpoint: revision.serving.knative.dev "" not found
````

**To Reproduce**

1. Install apisix with ingress and Gateway api enable (v1.5.0)
2. 
```
helm install openfunction openfunction/openfunction \
  -n openfunction \
  --set global.Contour.enabled=false
```
5.
```
kubectl apply -f - <<EOF
apiVersion: gateway.networking.k8s.io/v1alpha2
kind: GatewayClass
metadata:
  name: apisix-lb
spec:
  controllerName: apisix.apache.org/gateway-controller
EOF
```
6.
```
kubectl apply -f - <<EOF
apiVersion: networking.openfunction.io/v1alpha1
kind: Gateway
metadata:
  name: openfunction-gateway
  namespace: default
spec:
  domain: ofn.io
  clusterDomain: cluster.local
  hostTemplate: "{{.Name}}.{{.Namespace}}.{{.Domain}}"
  pathTemplate: "{{.Namespace}}/{{.Name}}"
  httpRouteLabelKey: "app.kubernetes.io/managed-by"
  gatewayDef:
    namespace: default
    gatewayClassName: apisix-lb
  gatewaySpec:
    listeners:
      - name: ofn-http-internal
        hostname: "*.cluster.local"
        protocol: HTTP
        port: 80
        allowedRoutes:
          namespaces:
            from: All
      - name: ofn-http-external
        hostname: "*.ofn.io"
        protocol: HTTP
        port: 80
        allowedRoutes:
          namespaces:
            from: All
EOF
```
7.
```
kubectl apply -f - <<EOF
apiVersion: core.openfunction.io/v1beta1
kind: Function
metadata:
  name: function-sample
spec:
  version: "v2.0.0"
  image: "localrepo/sample-go-func:v1"  
  imageCredentials:
    name: push-secret
  port: 8080 # default to 8080
  build:
    builder: openfunction/builder-go:latest
    env:
      FUNC_NAME: "HelloWorld"
      FUNC_CLEAR_SOURCE: "true"
    srcRepo:
      url: "https://github.com/OpenFunction/samples.git"
      sourceSubPath: "functions/knative/hello-world-go"
      revision: "main"
  serving:
    template:
      containers:
        - name: function # DO NOT change this
          imagePullPolicy: IfNotPresent 
    runtime: "knative"
  route:
    gatewayRef:
      name: openfunction-gateway
      namespace: default
    rules:
      - matches:
          - path:
              type: PathPrefix
              value: /v2/foo      
    hostnames:
    - "sample.ofn.io"      
EOF
```
7. See the error
```
export SERVICE_IP=$(kubectl get svc --namespace ingress-apisix apisix-gateway --template "{{ range (index .status.loadBalancer.ingress 0) }}{{.}}{{ end }}")
curl -i $SERVICE_IP/v2/foo/world -H 'Host: sample.ofn.io'

HTTP/1.1 404 Not Found
Content-Type: text/plain; charset=utf-8
Content-Length: 73
Connection: keep-alive
X-Content-Type-Options: nosniff
Date: Tue, 30 Aug 2022 21:41:02 GMT
Server: APISIX/2.15.0

Error getting active endpoint: revision.serving.knative.dev "" not found
```

**Screenshots**

We can see the service:

![Screen Shot 2022-08-30 at 5 44 57 PM](https://user-images.githubusercontent.com/25473327/187549276-d9600b08-df73-4ca1-83d4-f0fc309cd835.png)

**Comment by MBoui** [comment]
i also created the HTTPRoute

```
kubectl apply -f - <<EOF
apiVersion: gateway.networking.k8s.io/v1beta1
kind: HTTPRoute
metadata:
  name: hello-world-go
spec:
  parentRefs:
  - name: openfunction-gateway
    namespace: default 
  hostnames: ["sample.default.ofn.io", "sample.ofn.io"]
  rules:
    - matches:
      - path:
          type: PathPrefix
          value: /*
      filters:
        # replace tp the actual host of the service in knative
        - type: RequestHeaderModifier
          requestHeaderModifier:
            add:
            - name: Host
              value: serving-v7mwk-ksvc-bc7jv.default.example.com
      backendRefs:
        - name: serving-v7mwk-ksvc-bc7jv-v200
          weight: 1
          port: 80
EOF
```

**Comment by wrongerror** [comment]
Hi, @MBoui , This is because Knative currently only supports `Kourier`, `Istio` and `Contour` as network layers.

When using `Istio` /`Contour` as Gateway API implementation, OpenFuntion can share Istio/Contour with Knative.

When using a Gateway API implementation not supported by knative, you need to install a separate controller  for the network layer of Knative，see [install-a-networking-layer-for-knative](https://knative.dev/docs/install/yaml-install/serving/install-serving-with-yaml/#install-a-networking-layer). Otherwise the Knative service will not be ready because of `IngressNotConfigured`,

Thanks for your feedback, we will provide documentation support for switching to Gateway API implementations not supported by Knative.

**Comment by MBoui** [comment]
Hmm ok, so if i understand, Knative  (OpenFunction) don't support Apisix as not networking layer are available for Knative.
I am right?

**Comment by wrongerror** [comment]
Yeah，and we can install a separate controller(Kourier) for Knative , if this works,  i will update the documentation after testing with apisix-ingress-controller v1.5.0-rc.1.

**Comment by MBoui** [comment]
Ok so i followed the doc to install `Kourier`.
Now when i created a function, i see a new httproute created and then a new apisix route created also because of the `API Gateway` integration.

```
NAME              BUILDSTATE   SERVINGSTATE   BUILDER         SERVING         ADDRESS                                                   AGE
function-sample   Succeeded    Running        builder-z72zl   serving-lnql9   http://function-sample.default.svc.cluster.local/v2/foo   8m46s
```

When i try:
`curl -i $SERVICE_IP/v2/foo/world -H 'Host: sample.ofn.io'` i got the same error.

but when i try:
`curl -i $SERVICE_IP/v2/foo/world -H 'Host: function-sample.default.svc.cluster.local'`, i got:

```
curl -i $SERVICE_IP/v2/foo/world -H 'Host: function-sample.default.svc.cluster.local'
HTTP/1.1 404 Not Found
Content-Type: text/plain; charset=utf-8
Content-Length: 88
Connection: keep-alive
X-Content-Type-Options: nosniff
Date: Wed, 31 Aug 2022 15:12:04 GMT
Server: APISIX/2.15.0

Error getting active endpoint: revision.serving.knative.dev "function-sample" not found
```

which is new..

**Comment by MBoui** [comment]
> ```shell
> kubectl get httproute function-sample -oyaml
> ```

```
apiVersion: gateway.networking.k8s.io/v1beta1
kind: HTTPRoute
metadata:
  creationTimestamp: "2022-08-31T15:25:23Z"
  generation: 1
  labels:
    app.kubernetes.io/managed-by: default.openfunction-gateway
  name: function-sample
  namespace: default
  ownerReferences:
  - apiVersion: core.openfunction.io/v1beta1
    blockOwnerDeletion: true
    controller: true
    kind: Function
    name: function-sample
    uid: cf8f528d-b863-4059-8d6c-76eb9273ecb1
  resourceVersion: "568631"
  uid: 3322a73f-ac56-4eca-921e-d99a522c3228
spec:
  hostnames:
  - sample.ofn.io
  - function-sample.default.svc.cluster.local
  parentRefs:
  - group: gateway.networking.k8s.io
    kind: Gateway
    name: openfunction-gateway
    namespace: default
  rules:
  - backendRefs:
    - group: ""
      kind: Service
      name: serving-mkfsl-ksvc-96kgw-v200
      namespace: default
      port: 80
      weight: 1
    filters:
    - requestHeaderModifier:
        add:
        - name: Host
          value: serving-mkfsl-ksvc-96kgw-v200.default.svc.cluster.local
      type: RequestHeaderModifier
    matches:
    - path:
        type: PathPrefix
        value: /v2/foo
```

**Comment by MBoui** [comment]
> looks correct， the httproute you manually created for the function can be deleted.

Yeap, i did it.
But now i am getting a new error: `Error getting active endpoint: revision.serving.knative.dev "function-sample" not found`
Which i try to figure out why now.

**Comment by wrongerror** [comment]
yeah, I will try to create native k8s gateway & httproute and knative service directly without OpenFunction. If it still does not work, I think need to check how apisix implements Gateway API.

**Comment by wrongerror** [comment]
> Yeap, i did it. But now i am getting a new error: `Error getting active endpoint: revision.serving.knative.dev "function-sample" not found` Which i try to figure out why now.

I think that's why it doesn't work:
https://github.com/apache/apisix-ingress-controller/blob/1.5.0-rc1/pkg/ingress/gateway/translation/gateway_httproute.go#L63

apisix-ingress-controller currently does not handle filters，so Knative can't find the active endpoint based on the host in the headers.

**Comment by tao12345666333** [comment]
> apisix-ingress-controller currently does not handle filters

Yes. APISIX Ingress controller has not implemented this function yet.

**Comment by benjaminhuo** [comment]
> > apisix-ingress-controller currently does not handle filters
> 
> Yes. APISIX Ingress controller has not implemented this function yet.

@tao12345666333 Is there any plan to support this?

**Comment by tao12345666333** [comment]
Will be implemented soon, but since the development window for APISIX Ingress v1.5 has been closed, this feature will be carried over to v1.6

**Comment by MBoui** [comment]
Yeah, i saw that this morning by continuing my test. Thanks for your help, really appreciate. I will close the issue.
